### PR TITLE
[Block Executor] Fix halting logic

### PR DESCRIPTION
**Pull request by gelash**
Fix the issues stemming from https://github.com/aptos-labs/aptos-core/pull/11832 and beyond.

Previously, the errors were set during sequential commit preparation of the transactions. That allowed to set the first error encountered, as well as making sure errors were processed before non-error conditions for halting the scheduler / parallel execution (e.g. block gas limit, but also skip rest, or just having executed everything). However, even that logic was a bit flowed, since module publishing fallback can be observed non-deterministically (in terms of which transaction observes it), and also one of the motivations for above refactoring was to make error flows uniform, as some errors (such as delayed fields code invariants) can occur during execution (i.e. not in commit).

New design:
- abort / VMError in parallel also goes through the sequential fallback flow
- all errors in parallel execution still get ?-forwarded to the outside worker loop - each worker can arrive there with 1 error.
- we log all these errors at appropriate levels, but sequential fallback doesn't care what caused it
- error variants are simplified.

Passes tests (some adjustment due to error shortcut flows), added tests for halt in the schedule, as well as failpoint based test that ensures an error after halt still causes sequential fallback.

### CODE REVIEW

File: aptos-move/block-executor/src/errors.rs [hunk 1]
```diff
@@ -17,14 +17,39 @@ pub enum IntentionalFallbackToSequential {
...
+}
+
+pub(crate) fn log_panic_or(error: &PanicOr<IntentionalFallbackToSequential>, context: &str) {
+    match error {
+        PanicOr::Or(e) => match e {
+            IntentionalFallbackToSequential::ModulePathReadWrite => {
+                debug!("[Execution]: Module read & written {}", context);
+            },
+            IntentionalFallbackToSequential::ResourceGroupSerializationError(msg) => {
+                error!(
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/errors.rs:32]: should this be alert! ? (and below ones as well)

File: aptos-move/block-executor/src/errors.rs [hunk 2]
```diff
@@ -17,14 +17,39 @@ pub enum IntentionalFallbackToSequential {
...
+            },
+        },
+        PanicOr::CodeInvariantError(msg) => {
+            error!("[Execution]: CodeInvariantError({:?}), {}", msg, context);
+        },
+    }
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum BlockExecutionError<E> {
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/errors.rs:48]: we can leave for separate PR, but should we just have a different Error being returned by parallel and sequential execution, and removing this error type completely?

that way parallel execution cannot accidentally return BlockExecutionError::FatalVMError, which it never should

File: aptos-move/block-executor/src/executor.rs [hunk 3]
```diff
@@ -240,7 +240,12 @@ where
             },
             ExecutionStatus::Abort(err) => {
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/executor.rs:241]: where is regular transaction abort and commit onchain handled? is that in ExecutionStatus::Success ?

File: aptos-move/block-executor/src/executor.rs [hunk 4]
```diff
@@ -240,7 +240,12 @@ where
...
                 // Record the status indicating abort.
-                ExecutionStatus::Abort(BlockExecutionError::FatalVMError((err, idx_to_execute)))
+                ExecutionStatus::Abort(BlockExecutionError::FallbackToSequential(PanicOr::Or(
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/executor.rs:243]: why do we just not return exception here? is any of the work below useful, if we are going to fallback to sequential anyways?

File: aptos-move/block-executor/src/scheduler.rs [hunk 5]
```diff
@@ -272,7 +281,7 @@ pub struct Scheduler {
     /// Shared marker that is set when a thread detects that all txns can be committed.
     done_marker: CachePadded<AtomicBool>,
 
-    has_halted: CachePadded<AtomicBool>,
+    halted_status: CachePadded<Mutex<HaltedStatus>>,
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/scheduler.rs:284]: will this have performance issues? 

alternativelly, you could have an enum with incremental values for each, that you store in AtomicU8?

File: aptos-move/block-executor/src/executor.rs [hunk 6]
```diff
@@ -573,18 +578,13 @@ where
...
+            // 3) A transaction triggered an Abort (unrecoverable VM error).
             // We cover cases 1 and 2 here
             maybe_code_error.and_then(|_| last_input_output.module_rw_intersection_ok())?;
-
             // Next, we handle 3, an abort / an unrecoverable VM error.
-            if let Some(err) = last_input_output.aborted_execution_status(txn_idx) {
-                assert!(matches!(err, BlockExecutionError::FatalVMError(_)));
-                return Err(err);
-            }
+            last_input_output.aborted_execution_status(txn_idx)?;
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/executor.rs:585]: this should now be name:
handle_aborted_execution_status or check_aborted_execution_status , as it doesn't return the status?

File: aptos-move/block-executor/src/errors.rs (hunk 2 above)
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:48]: good point, I added unreachables, but that's much cleaner. Let's do in a separate PR though, might add too much noise here.

File: aptos-move/block-executor/src/executor.rs (hunk 3 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:241]: Yes, as far as I know even that sort of failure is part of Success status.

File: aptos-move/block-executor/src/executor.rs (hunk 4 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:243]: This might still be speculative

Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:243]: but it shouldn't happen, so even if there was a speculative Abort, and got cleaned up, we would want to know about it. Especially now with the fallback that's absolutely the best way to handle it, and would also simplify commit flow. Great, I'll try that.

File: aptos-move/block-executor/src/scheduler.rs (hunk 5 above)
Comment (gelash) [reply aptos-move/block-executor/src/scheduler.rs:284]: Only called at most once by each worker, that's why I didn't optimize too much.

File: aptos-move/block-executor/src/executor.rs (hunk 6 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:585]: Check makes sense, and if we fast-forward the abort, will fit even better.

File: aptos-move/block-executor/src/errors.rs (hunk 2 above)
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:48]: I can try to make the change and see

File: aptos-move/block-executor/src/errors.rs (hunk 1 above)
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:32]: probably makes sense. The resource group error is alerting when it's created too, but being defensive with this makes sense to me.

File: aptos-move/block-executor/src/executor.rs [hunk 7]
```diff
@@ -1458,25 +1460,7 @@ where
...
-                        );
-                    },
-                    PanicOr::CodeInvariantError(msg) => {
-                        error!(
-                            "[Execution]: CodeInvariantError({:?}), sequential fallback",
-                            msg
-                        );
-                    },
-                };
+                log_panic_or(e, "sequential_fallback");
```
Comment (gelash) [inline aptos-move/block-executor/src/executor.rs:1463]: We can just debug here that fallback happened for reason e.

File: aptos-move/block-executor/src/errors.rs (hunk 2 above)
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:50]: made some steps in that direction.

**Review by gelash — COMMENTED** [review]
addressing comments.

**Review by igor-aptos — APPROVED** [review]

File: aptos-move/block-executor/src/unit_tests/mod.rs [hunk 8]
```diff
@@ -28,11 +30,64 @@ use aptos_types::{
...
+        DeltaDataView<KeyType<u32>>,
+        NoOpTransactionCommitHook<MockOutput<KeyType<u32>, MockEvent>, usize>,
+        ExecutableTestType,
+    >::new(
+        BlockExecutorConfig::new_no_block_limit(num_cpus::get()),
+        executor_thread_pool,
+        None,
+    );
+
+    let scenario = FailScenario::setup();
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/unit_tests/mod.rs:75]: Didn't know it can be done this way, neat

File: aptos-move/block-executor/src/errors.rs [hunk 9]
```diff
@@ -17,14 +18,40 @@ pub enum IntentionalFallbackToSequential {
     /// We defensively check resource group serialization error in the commit phase.
     /// TODO: should trigger invariant violation in the transaction itself.
     ResourceGroupSerializationError(String),
+    /// Execution of a thread yields a non-recoverable error from the VM. It it happens
+    /// during parallel execution, an error will be logged and sequential fallback taken.
+    FatalVMError(String),
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/errors.rs:23]: Do we need these variants? We can have just an `IntentionalFallbackToSequential` and the caller will log whether it is a module read-write error, resource group, etc.? Or am I missing something?

File: aptos-move/block-executor/src/executor.rs [hunk 10]
```diff
@@ -538,53 +545,39 @@ where
                         finalized_groups.push((group_key, metadata_op, finalized_group));
                     },
                     Err(err) => {
-                        maybe_code_error = Err(err.into());
-                        break;
+                        return Err(err.into());
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:548]: So all error condition here trigger a fallback, right? And therefore we can just return here, so we exit the worker loop and halt the block, falling to sequential?

File: aptos-move/block-executor/src/proptest_types/baseline.rs [hunk 11]
```diff
@@ -362,12 +362,16 @@ impl<K: Debug + Hash + Clone + Eq> BaselineOutput<K> {
...
                 });
             },
-            Err(BlockExecutionError::FatalVMError((idx, executor_idx))) => {
-                assert_eq!(*idx, *executor_idx as usize);
+            Err(BlockExecutionError::FatalVMError(idx)) => {
                 assert_matches!(&self.status, BaselineStatus::Aborted);
                 assert_eq!(*idx, self.read_values.len());
                 assert_eq!(*idx, self.resolved_deltas.len());
             },
+            Err(BlockExecutionError::FallbackToSequential(PanicOr::Or(
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/proptest_types/baseline.rs:370]: Why we have an extra match case here, when there is `Err(BlockExecutionError::FallbackToSequential(e))` below?

File: aptos-move/block-executor/src/scheduler.rs [hunk 12]
```diff
@@ -600,8 +601,53 @@ impl Scheduler {
...
+    /// 1. There is a module publishing txn that has read/write intersection with any txns
+    ///    even during speculative execution.
+    /// 2. There is a resource group serialization error.
+    /// 3. There is a txn with VM execution status Abort.
+    /// 4. There is a txn with VM execution status SkipRest.
+    /// 5. The committed txns have exceeded the PER_BLOCK_GAS_LIMIT.
+    /// 6. All transactions have been executed.
+    ///
+    /// For scenarios 1, 2 & 3, the output of the block execution will be an error, leading
+    /// to a fallback with sequential execution. For scenarios 3, 4 & 5, execution outputs
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/scheduler.rs:616]: typo: `4, 5, 6`?

File: aptos-move/block-executor/src/executor.rs [hunk 13]
```diff
@@ -1458,25 +1449,7 @@ where
...
-                        );
-                    },
-                    PanicOr::CodeInvariantError(msg) => {
-                        error!(
-                            "[Execution]: CodeInvariantError({:?}), sequential fallback",
-                            msg
-                        );
-                    },
-                };
+                debug!("Sequential_fallback occurred, reason: {:?}", e);
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:1452]: Doesn't `log_panic_or` already logs `e` somewhere else?

File: aptos-move/block-executor/src/executor.rs [hunk 14]
```diff
@@ -1091,20 +1087,14 @@ where
...
-                                let block_limit_processor = shared_commit_state.acquire();
-                                info!("Block execution was aborted due to {:?}", inner_err);
-                                block_limit_processor.finish_parallel_update_counters_and_log_info(
-                                    txn_idx + 1,
-                                    scheduler.num_txns(),
-                                );
-                            }
-
-                            *shared_maybe_error.acquire() = Err(e);
+                        log_panic_or(&e, "[BlockSTM] Parallel execution worker loop halt_err");
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:1090]: Why does log come before `if scheduler.halt_err()`?

File: aptos-move/block-executor/src/executor.rs [hunk 15]
```diff
@@ -601,11 +594,14 @@ where
...
+                if scheduler.halt_ok() {
                     block_limit_processor.finish_parallel_update_counters_and_log_info(
                         txn_idx + 1,
                         scheduler.num_txns(),
                     );
+                    fail_point!("commit-all-halt-err", |_| Err(PanicOr::Or(
+                        IntentionalFallbackToSequential::ResourceGroupSerializationError("".into())
+                    )));
                 }
                 break;
```
Comment (georgemitenkov) [inline aptos-move/block-executor/src/executor.rs:606]: I know we have Ok(()) after while commit loop terminates, but this technically should be a return?

File: aptos-move/block-executor/src/executor.rs [hunk 16]
```diff
@@ -538,53 +545,39 @@ where
...
-                        break;
+                        return Err(err.into());
                     },
                 }
             }
 
-            if maybe_code_error.is_ok() {
-                if let Some(group_reads_needing_delayed_field_exchange) =
-                    last_input_output.group_reads_needing_delayed_field_exchange(txn_idx)
+            if let Some(group_reads_needing_delayed_field_exchange) =
```
Comment (ziaptos) [inline aptos-move/block-executor/src/executor.rs:553]: the "needing" is really poking my eyes :) -> "require"

File: aptos-move/block-executor/src/errors.rs [hunk 17]
```diff
@@ -17,14 +18,40 @@ pub enum IntentionalFallbackToSequential {
...
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum BlockExecutionError<E> {
     FallbackToSequential(PanicOr<IntentionalFallbackToSequential>),
-    /// Execution of a thread yields a non-recoverable error from the VM. Such an error will be propagated
-    /// back to the caller (leading to the block execution getting aborted).
-    FatalVMError((E, TxnIndex)),
+    /// If the unrecoverable VM error occurs during sequential execution (e.g. fallback),
+    /// the error is propagated back to the caller (block execution is aborted).
+    FatalVMError(E),
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:54]: you mean call the one in fallback and the one outside differently?

File: aptos-move/block-executor/src/executor.rs (hunk 10 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:548]: yes, we can forward all errors now.

File: aptos-move/block-executor/src/proptest_types/baseline.rs (hunk 11 above)
Comment (ziaptos) [reply aptos-move/block-executor/src/proptest_types/baseline.rs:370]: @georgemitenkov Looks like we have it b/c we only expect the PanicOr variant of e.
@gelash I'd re-do this match (to have one enum variant, and the redo on the matched e - or use match!). This is for performance reasons (to allow for a jump table).

Comment (gelash) [reply aptos-move/block-executor/src/proptest_types/baseline.rs:370]: because this one is now okay, but the rest is unreachable. I can add explicit variants (it was in a test)

File: aptos-move/block-executor/src/executor.rs (hunk 14 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:1090]: because we want to log all errors, but only one of them will trigger the fallback

File: aptos-move/block-executor/src/scheduler.rs [hunk 18]
```diff
@@ -49,6 +49,16 @@ impl ArmedLock {
     }
 }
 
+#[derive(Debug)]
+enum HaltedStatus {
```
Comment (ziaptos) [inline aptos-move/block-executor/src/scheduler.rs:53]: Strong opinion here:

The "halted" term has lost it's meaning long time ago, and is very, very, confusing. Why not rename it to just ExecutionStatus, with something like "Executing", "DoneProcessing", "Aborting". Or something similar.

File: aptos-move/block-executor/src/executor.rs (hunk 14 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:1090]: (I wonder if I should add transaction index everywhere, that might be helpful)

File: aptos-move/block-executor/src/scheduler.rs [hunk 19]
```diff
@@ -600,8 +601,53 @@ impl Scheduler {
                 self.halt_transaction_execution(txn_idx);
             }
         }
+    }
+
+    /// Currently, the reasons for halting the scheduler are as follows:
+    /// 1. There is a module publishing txn that has read/write intersection with any txns
```
Comment (ziaptos) [inline aptos-move/block-executor/src/scheduler.rs:607]: Now when we do have a mutex protected status, we can actually have all of these be a separate status.

**Review by ziaptos — CHANGES_REQUESTED** [review]

File: aptos-move/block-executor/src/errors.rs [hunk 20]
```diff
@@ -17,14 +18,40 @@ pub enum IntentionalFallbackToSequential {
     /// We defensively check resource group serialization error in the commit phase.
     /// TODO: should trigger invariant violation in the transaction itself.
     ResourceGroupSerializationError(String),
+    /// Execution of a thread yields a non-recoverable error from the VM. It it happens
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:21]: it's not about logging, the one here is when it happens during parallel execution (even speculative, just added that to the comment). The one below is when it happens during sequential / fallback, then we have to pass it through. I'll rename the names as @georgemitenkov suggested, lmk what else.

File: aptos-move/block-executor/src/errors.rs (hunk 9 above)
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:23]: Fair enough from returning from parallel point of view, but we still have to propagate the errors to the main worker loop?

File: aptos-move/block-executor/src/errors.rs [hunk 21]
```diff
@@ -17,14 +18,40 @@ pub enum IntentionalFallbackToSequential {
...
+    /// during parallel execution, an error will be logged and sequential fallback taken.
+    FatalVMError(String),
+}
+
+pub(crate) fn log_panic_or(error: &PanicOr<IntentionalFallbackToSequential>, context: &str) {
```
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:26]: However, I still like it this way: first it removes logic from executor, which is already bloated. Second, it isolates the handling logic for different variants - some are expected and need debug, some need alerts. Could also be used in some other context.

File: aptos-move/block-executor/src/errors.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:54]: @ziaptos if you read the comments instead of asking to delete those, maybe you'd know xDD but I don't trust my explanation skills.

[the remark this answers is not in the document]
Comment (ziaptos) [reply aptos-move/block-executor/src/errors.rs:54]: Oh, so the error is extracted from IntentionalFallbackToSequential and put into this special case?

[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:54]: kidding, it wasn't understandable - made it to the point.

File: aptos-move/block-executor/src/scheduler.rs (hunk 18 above)
Comment (gelash) [reply aptos-move/block-executor/src/scheduler.rs:53]: This PR is a bug fix. The thing was called halt.
ExecutionStatus also already exists.

File: aptos-move/block-executor/src/executor.rs (hunk 16 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:553]: I'm not changing it in this PR, that change would touch 30 files.

Comment (gelash) [reply aptos-move/block-executor/src/executor.rs:553]: if @igor-aptos though, I can send a PR to bulk replace, not lazy, not a problem :)

File: aptos-move/block-executor/src/scheduler.rs (hunk 19 above)
Comment (ziaptos) [reply aptos-move/block-executor/src/scheduler.rs:607]: I think this one is useful - to understand and verify code thats happening here

File: aptos-move/block-executor/src/scheduler.rs (hunk 18 above)
Comment (ziaptos) [reply aptos-move/block-executor/src/scheduler.rs:53]: @gelash I'm standing by my opinion:

1. I just suggested some namings (ExecutionStage, ExecutionState are some other alternatives).
2. This PR introduces HaltedOk and HaltedErr (two very confusing terms). being a bug fix is not a reason not to give more relevant naming here.

**Review by ziaptos — CHANGES_REQUESTED** [review]

File: aptos-move/block-executor/src/scheduler.rs (hunk 18 above)
Comment (gelash) [reply aptos-move/block-executor/src/scheduler.rs:53]: I think I can revert all this - since we don't care which error causes fallback - and maybe also just return empty error. Working on it right now.

File: aptos-move/block-executor/src/executor.rs (hunk 16 above)
Comment (igor-aptos) [reply aptos-move/block-executor/src/executor.rs:553]: I'm fine with group_reads_requiring_delayed_field_exchange, but for my English- what is wrong with "needing"? Sounds like correct English to me ? :)

Comment (ziaptos) [reply aptos-move/block-executor/src/executor.rs:553]: it's "needling" my eyes :) Dunno, just sounds off (and require is kind of standard)

File: aptos-move/block-executor/src/errors.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/block-executor/src/errors.rs:54]: I think parralel should return PanicOr<ModulePathReadWrite>.

I.e. all others are equivalent to panics and handled the same

[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:54]: If parallel and sequential return different types, fallback handling gets messy. I'm sure there is a good way, but I didn't find it there. Instead, I made IntentionalFallbackToSequential not have any subtypes. Sending now.

**Comment by gelash** [comment]
Please have another look. I tried my best to make it fit all comments.

File: aptos-move/block-executor/src/errors.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/block-executor/src/errors.rs:32]: I don't get it - errrors from parralel are never the ones returned, they are internal (as we always fall back)

We dont need them to match?

File: aptos-move/block-executor/src/errors.rs (hunk 2 above)
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:28]: This is now super easy to check because it literally creates an error at return.

File: aptos-move/block-executor/src/errors.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:32]: it dispatches on concurrency level, gets a ret, and then checks that ret for fallback.
coming soon we can fallback from sequential to sequential with pre-serialization also, so the above is actually useful. 
I can get the error and map it though, let me do that.

[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:32]: ah but all the tests.. I can do it right after.

[the remark this answers is not in the document]
Comment (igor-aptos) [reply aptos-move/block-executor/src/errors.rs:32]: Or you mean :;execute(transaction), hm

File: aptos-move/block-executor/src/errors.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs]: this is all gone.

File: aptos-move/block-executor/src/errors.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs]: this is in the constructor. we need to make sure these are logged.

File: aptos-move/block-executor/src/scheduler.rs (hunk 18 above)
Comment (gelash) [reply aptos-move/block-executor/src/scheduler.rs]: this is all gone back to the way you had it.

File: aptos-move/block-executor/src/errors.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (gelash) [reply aptos-move/block-executor/src/errors.rs:32]: done!

File: aptos-move/block-executor/src/executor.rs [hunk 22]
```diff
@@ -1394,24 +1399,21 @@ where
...
                     }
-                    // Record the status indicating abort.
-                    return Err(BlockExecutionError::FatalVMError((err, idx as TxnIndex)));
+                    alert!("Fatal VM error by transaction {}", idx as TxnIndex);
+                    // Record the status indicating the unrecoverable VM failure.
+                    return Err(BlockExecutionError::FatalVMError(err));
+                },
+                ExecutionStatus::DelayedFieldsCodeInvariantError(msg) => {
+                    return Err(BlockExecutionError::FallbackToSequential(
+                        PanicOr::CodeInvariantError(msg),
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/executor.rs:1408]: we generally should call `code_invariant_error(msg)` instead of creating directly (if you need you can add .into(), if it doesn't get casted automatically from Panic to PanicOr)

File: aptos-move/block-executor/src/txn_last_input_output.rs [hunk 23]
```diff
@@ -211,33 +196,35 @@ impl<T: Transaction, O: TransactionOutput<Txn = T>, E: Debug + Send + Clone>
...
+                    Err(code_invariant_error("Speculative error status cannot be committed").into())
+                },
+                ExecutionStatus::DelayedFieldsCodeInvariantError(_) => Err(code_invariant_error(
+                    "Delayed field invariant error cannot be committed",
+                )
+                .into()),
+            }
         } else {
+            // TODO: check carefully if such a case is possible due to early halts.
+            debug!("Possible to not find committed output (halted?)");
```
Comment (igor-aptos) [inline aptos-move/block-executor/src/txn_last_input_output.rs:223]: warn instead of debug, so we actually see it?

also do you want to quality TODO?

**Review by igor-aptos — APPROVED** [review]

File: aptos-move/block-executor/src/txn_last_input_output.rs (hunk 23 above)
Comment (gelash) [reply aptos-move/block-executor/src/txn_last_input_output.rs]: Thanks, good point!
I will go over general block executor TODOs after all the refactors and classify them.

File: aptos-move/block-executor/src/executor.rs (hunk 22 above)
Comment (gelash) [reply aptos-move/block-executor/src/executor.rs]: braindead editing, thanks for catching

Comment (gelash) [reply aptos-move/block-executor/src/executor.rs]: it's so dumb that we can't make types uninstantiable - only way is via visibility and it doesn't work if you need to include the type elsewhere.

**Review by ziaptos — APPROVED** [review]
